### PR TITLE
Feature/sign in functionality

### PR DESCRIPTION
**Pull request by eugenialcaraz**
#### Trello board reference:

* [6](Add log in functionality)

---

#### Description:

* Add connection to API for sign in
* Add input validations for empty fields
* Add redirection if sign in is done correctly
* Add error messages
* Add .env variables

---

#### Reviewers:

* @nicmalegre 
* @jpirazusta 
* @tobicorradi 
* @diego317 

---

#### Notes:

* Sign up form will be updated in a separated ticket, only the new props were added so the branch doesn't break.
* Styles of main page are not ready yet, they are temporary.
* .env file tracked for now, so you can access it if your need it
* The accepted user for now is the following:

```
        email: juan@ferreira.com
        password": password
```

---

#### Preview:

image

image

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,2 @@
+VITE_BASE_URL = https://target-api-induction-v2.herokuapp.com/api/v1/users
```
Comment (nicmalegre) [inline .env]: I know in this project there isn’t too much risk but maybe you could add this .env to gitignore and only add to the repo a env.example with fake variables

File: .env [hunk 2]
```diff
@@ -0,0 +1,2 @@
...
+VITE_SIGN_IN =/sign_in
```
Comment (jpirazusta) [inline .env]: I don't think you should include the `/users` in the api url (not all routes have that path) and add an environment variable for the `/sign in` route. Also, remember to add this file to gitignore if you then add some key

File: src/components/common/forms/SignInForm.tsx [hunk 3]
```diff
@@ -1,31 +1,63 @@
...
+  const onSubmit = async (data: object) => {
+    const response = await signInRequest({
+      user: data,
+    });
+    if (typeof response === "object") {
+      setLocalStorage("user", response);
+      navigate(urlFormat(Pages.Main));
+    } else {
+      setError("serverError", { type: "custom", message: response });
+    }
```
Comment (nicmalegre) [inline src/components/common/forms/SignInForm.tsx]: I'm not sure to handle success/error based on response type, maybe you could add a try / catch in `onSubmit` or maybe some status inside response to handle if the request succeed ?

File: src/components/common/forms/Forms.module.css [hunk 4]
```diff
@@ -10,3 +11,19 @@
...
 }
+
+.form > span {
+  font-size: 1.2rem;
+  line-height: 1.4rem;
+  letter-spacing: 0.18rem;
+  margin-bottom: 0.8rem;
+  color: var(--error);
+  overflow: hidden;
+  height: 0rem;
```
Comment (jpirazusta) [inline src/components/common/forms/Forms.module.css]:
```suggestion
  height: 0;
```
😅

File: src/components/common/forms/SignInForm.tsx [hunk 5]
```diff
@@ -1,31 +1,63 @@
...
   };
 
+  const isFormValid = Object.keys(errors).length === 0;
```
Comment (jpirazusta) [inline src/components/common/forms/SignInForm.tsx]: Same suggestion as in the other file

File: src/components/common/forms/SignInForm.tsx [hunk 6]
```diff
@@ -1,31 +1,63 @@
...
+            "Both fields are required"}
+      </span>
       <Input
         label="email"
         type="email"
-        stylesName="signIn"
+        stylesName={isFormValid ? "signIn" : "error"}
         register={register}
         required
+        onChange={() => clearErrors()}
```
Comment (jpirazusta) [inline src/components/common/forms/SignInForm.tsx]: Should it clear all errors, or just this field's errors?

File: src/components/common/forms/SignInForm.tsx [hunk 7]
```diff
@@ -1,31 +1,63 @@
...
       />
       <Input
         label="password"
         type="password"
-        stylesName="signIn"
+        stylesName={isFormValid ? "signIn" : "error"}
         register={register}
         required
+        onChange={() => clearErrors()}
```
Comment (jpirazusta) [inline src/components/common/forms/SignInForm.tsx]: Same question

**Review by nicmalegre — APPROVED** [review]

File: src/utils/localStorage.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    return value && JSON.parse(value);
+  }
+};
+
+export const setLocalStorage = (key: string, value: object) => {
+  localStorage.setItem(key.toLowerCase(), JSON.stringify(value));
+};
+
+export const clearLocalStorage = () => {
+  localStorage && localStorage.clear();
```
Comment (jpirazusta) [inline src/utils/localStorage.tsx]:
```suggestion
  localStorage?.clear();
```

**Review by jpirazusta — APPROVED** [review]

File: src/utils/localStorage.tsx (hunk 8 above)
Comment (eugenialcaraz) [reply src/utils/localStorage.tsx]: Super hacker, love it

File: src/components/common/forms/SignInForm.tsx (hunk 7 above)
Comment (eugenialcaraz) [reply src/components/common/forms/SignInForm.tsx]: Here I also want to clear just the custom error `"serverError"`, I have updated it too!

File: src/components/common/forms/SignInForm.tsx (hunk 6 above)
Comment (eugenialcaraz) [reply src/components/common/forms/SignInForm.tsx]: I want to clear just the custom error `"serverError"`, I have updated it!

File: src/components/common/forms/SignInForm.tsx (hunk 5 above)
Comment (eugenialcaraz) [reply src/components/common/forms/SignInForm.tsx]: Updated!

File: src/components/common/forms/Forms.module.css (hunk 4 above)
Comment (eugenialcaraz) [reply src/components/common/forms/Forms.module.css]: 😅 Fixed!

File: .env (hunk 2 above)
Comment (eugenialcaraz) [reply .env]: Perfect, I've updated them accordingly. And also added the `.env` to git ignore, only left an `.env.example`

File: .env (hunk 1 above)
Comment (eugenialcaraz) [reply .env]: Perfect, I've added `env.example` and `.env` is ignored now

File: src/components/common/forms/SignInForm.tsx (hunk 3 above)
Comment (eugenialcaraz) [reply src/components/common/forms/SignInForm.tsx]: Perfect with try/catch looks way better 

```suggestion
   const onSubmit = async (data: object) => {
    try {
      const response = await signInRequest({
        user: data,
      });
      setLocalStorage("user", response);
      navigate(urlFormat(Pages.Main));
    } catch (error) {
      setError("serverError", {
        type: "custom",
        message: (error as Error).message,
      });
    }
  };
```

**Pull request merged**